autonomous_software_pkg/src/map_publisher_2.py

```diff
@@ -0,0 +1,210 @@
+
+            root = tree.getroot()
+
+            for p in root.Document.Placemark:
+
+                component = MapComponent(id=total_id)
+                total_id += 1
+                component.list_coordinates = []
+
+                # Classify type of component using its name
+                name = p.name.text
+
+                if "obstacle" in name:
+                    component.type = PlacemarkType.Obstacle
+
+                elif "contour" in name:
+                    component.type = PlacemarkType.Contour
+
+                elif "line" in name:
+                    component.type = PlacemarkType.Lane
+
+                # Extract coordinates from component (is different depending on geometry of component)
+                if hasattr(p, "Polygon"):
+                    coordinates = p.Polygon.outerBoundaryIs.LinearRing.coordinates
+
+                if hasattr(p, "LineString"):
+                    coordinates = p.LineString.coordinates
+
+                # Convert to string
+                coordinates_string = str(coordinates)
+
+                # Go through coordinates, convert them to utm and add them to component object
+                for x_y in coordinates_string.split():
+
+                    lon = float((x_y.split(","))[0])
+                    lat = float((x_y.split(","))[1])
+
+                    utm_values = utm.from_latlon(lat, lon)
+                    x_utm, y_utm = utm_values[0], utm_values[1]
+
+                    component.list_coordinates.append((x_utm, y_utm))
+
+                # Add map component to list of map components
+                list_components.append(component)
+
+        return list_components
+
+    def publish_map(self, map):
+        """
+        Function to publish a MarkerArray topic, containing Markers each representing a portion of the map
+        """
+
+        markerArray = MarkerArray()
+
+        id = 0
+
+        for component in map:
+
+            marker = Marker()
+            marker.header.frame_id = "world"
+            marker.type = marker.LINE_STRIP
+            marker.action = marker.ADD
+            marker.scale.x = 0.2
+            marker.scale.y = 0.2
+            marker.scale.z = 0.2
+            marker.color.a = 1.0
+
+            marker.pose.orientation.w = 1.0
+            marker.pose.position.x = 0
+            marker.pose.position.y = 0
+            marker.pose.position.z = 0
+
+            if component.type == PlacemarkType.Obstacle:
+                marker.color.r = 0.89
+                marker.color.g = 0.05
+                marker.color.b = 0.19
+
+            elif component.type == PlacemarkType.Lane:
+                marker.color.r = 0.08
+                marker.color.g = 0.63
+                marker.color.b = 0.09
+
+            elif component.type == PlacemarkType.Contour:
+                marker.color.r = 0.3
+                marker.color.g = 0.3
+                marker.color.b = 0.3
+
+            marker.id = id
+            id += 1
```

Comment:
Same as above, you can use `enumerate` to not have to keep track of the indices by hand.
Edit: in fact, you set an attribute called `id` when creating the element. Shouldn't we use it here instead of creating a new one?

Reply:
Done